### PR TITLE
chore: add support for building on jetson device

### PR DESCRIPTION
**Pull request by yusiwen**
The current building flow will fail on Jetson, because the CTranslate2 binding turns the "sse4.1" flag on while it's not available on Jetson.

I've added a function to check whether it's building on the NVIDIA Jetson device or not. If true, it will turn on CUDA and CUDNN flags of CTranslate2.

I've tested building on Intel CPU, AMD CPU, Jetson Orin Nano, Jetson Orin AGX.

### CODE REVIEW

File: crates/ctranslate2-bindings/build.rs [hunk 1]
```diff
@@ -37,9 +52,21 @@ fn link_static() -> PathBuf {
...
+               .define("WITH_CUDNN", "ON")
+               .cxxflag("-mcpu=native")
+        } else if cfg!(target_feature = "sse4.1") {
+            config
+                .define("WITH_OPENBLAS", "ON")
+                .cxxflag("-msse4.1")
+        } else {
+            config
+                .define("WITH_CUDA", "ON")
+        }
```
Comment (wsxiaoys) [inline crates/ctranslate2-bindings/build.rs:69]: Could you refactor above block with following branching structures?

```rust
if cfg!(target_os = "linux") {
  ...
  if cfg!(target_feature = "sse4.1") {
    ...
  }

  if cfg!(feature = "link_static_cuda") {
    // turn on cuda
    ...
    // if needed, branch here on aarch64 / x86
  } else {
    // Turn on openblas
    ...
  }
} else if cfg!(target_os = "macos") {
  ...
}
```

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: crates/ctranslate2-bindings/build.rs (hunk 1 above)
Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs:69]: Copy that! I will try to refactor it ASAP

Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs]: @wsxiaoys I've refactored as you requested. Please take a look at this revision

File: crates/ctranslate2-bindings/build.rs [hunk 2]
```diff
@@ -54,26 +54,28 @@ fn link_static() -> PathBuf {
...
-                .define("WITH_OPENBLAS", "ON")
-                .cxxflag("-msse4.1")
+        if cfg!(target_feature = "sse4.1") {
+            config.cxxflag("-msse4.1");
+        }
+        
+        if cfg!(feature = "link_static_cuda") {
+            config.define("WITH_CUDA", "ON");
+            
+            if cfg!(target_arch = "aarch64") && check_jetson() {
```
Comment (wsxiaoys) [inline crates/ctranslate2-bindings/build.rs:64]:
```suggestion
            if cfg!(target_arch = "aarch64") {
```

check_jetson is no longer needed as you'll need to turn on link_static_cuda manually```

File: crates/ctranslate2-bindings/build.rs [hunk 3]
```diff
@@ -54,26 +54,28 @@ fn link_static() -> PathBuf {
...
+                config
+                   .define("WITH_CUDNN", "ON")
```
Comment (wsxiaoys) [inline crates/ctranslate2-bindings/build.rs:66]: this should be enabled no matter if aarch64 / x86?

**Review by wsxiaoys — CHANGES_REQUESTED** [review]

File: crates/ctranslate2-bindings/build.rs (hunk 2 above)
Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs:64]: OK, got it.

File: crates/ctranslate2-bindings/build.rs (hunk 3 above)
Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs:66]: I found on Jetson cuDNN comes with JetPack by default. But on Linux (e.g., Ubuntu), cuDNN needs to be installed separately with CUDA. Does Tabby use cuDNN? If not, I'll remove it

Comment (wsxiaoys) [reply crates/ctranslate2-bindings/build.rs:66]: Should be good to align with ctranslate2's CUDA compilation configuration

https://github.com/OpenNMT/CTranslate2/blob/master/python/tools/prepare_build_environment_linux.sh#L56C1-L57C1

Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs:66]: OK, I got it.

Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs:66]: @wsxiaoys `CUDA_DYNAMIC_LOADING=ON` in the script given by you probably conflicts with the literal meaning of `link_static_cuda`

Comment (wsxiaoys) [reply crates/ctranslate2-bindings/build.rs:66]: I was specifically mentioning the `WITH_CUDNN=ON` in ctranslate2's cuda config

Comment (yusiwen) [reply crates/ctranslate2-bindings/build.rs]: @wsxiaoys How about this version. I've deleted unused code.

**Review by wsxiaoys — APPROVED** [review]
LGTM and running CIs

**Comment by yusiwen** [comment]
> LGTM and running CIs

Oops, seems that there is a lint problem. I'm pretty new to Rust, where should I have to look at?

**Comment by wsxiaoys** [comment]
You could run `cargo +nightly fmt` to fix the lint

**Comment by yusiwen** [comment]
Ok, I'm digging into it.

**Comment by yusiwen** [comment]
@wsxiaoys Wierd, My `cargo +nightly fmt --check` shows everything is fine. Do I miss something?

**Comment by yusiwen** [comment]
I've pushed the result of `cargo fmt`. The CI log doesn't show much of information about why it complains my code. How about this version?

**Comment by wsxiaoys** [comment]
> @wsxiaoys I've pushed the result of `cargo fmt`. The CI log doesn't show much of information about why it complains my code. How about this version?

image

This is screenshot from link https://github.com/TabbyML/tabby/actions/runs/6623245595/job/17990064469?pr=620

**Comment by yusiwen** [comment]
@wsxiaoys Yes, I've studied that log, cannot find any information why the code failed to lint. The error just shows "Process completed with exit code 1." at somewhere around line 42.

**Comment by yusiwen** [comment]
@wsxiaoys By the way, while still on the subject of NVIDIA Jetson. Should I add an additional Dockerfile for building images runs on Jeton devices? It will be named as `Dockerfile.jetson`, sounds good?

**Pull request merged**